tracks/julia/exercises/word-count/mentoring.md

```diff
@@ -121,4 +121,78 @@ function wordcount(sentence)
 end
 ```
 
+** Without Regex ** 
+
+This solution doesn't use a regex at all. It uses the default split behavior (which checks for `isspace` characters) and then strips punctuation characters left over on the "outside" of words. Filtering hoas to come after stripping in case the strip operation itself left an empty string.
+```julia
+function wordcount(sentence)
+    words = strip.(ispunct, split(lowercase(sentence)))
+    filter!(!isempty, words)
+    d = Dict{eltype(words), Int}()
+    for w in words
+        d[w] = get(d, w, 0) + 1
+    end
+    d
+end
+```
+It's possible to create the dictionary more efficiently with the `countmap` function from `StatsBase` (which only hashes the key once per check). With countmap, you can code-golf the above into:
+```julia
+using StatsBase: countmap
+
+wordcount_withstatsbase(sentence) = countmap(filter!(!isempty, strip.(ispunct, split(lowercase(sentence)))))
+```
+
+
+** Totally manual ** 
+
+This method iterates through the string one character at a time and switches behavior at word boundaries (when `iswordchar` changes from true to false or false to true). The dictionary has `SubString` keys in this version, which are like the string equivalent of Array `views`. It is about 20% faster than the StatsBase version on small inputs like those in the tests (performance is almost identicaly for inputs without thousands of characters). It's about 1.5x-2x faster than the next-fastest regex version, likely owing to the smaller number of allocations.
+```julia
+function wordcount(input)
+    input = lowercase(input)
+    d = Dict{SubString{typeof(input)}, Int}()
+    iswordchar(c) = !ispunct(c) && !isspace(c)
+
+    in_word = false
+    j = 0
+    for i in eachindex(input)
+        c = input[i]
+
+        if i == lastindex(input)
```

Comment:
You mentioned not liking this. Here's a `do while` loop that moves the lastindex check all the way to the end:

```julia
function wordcount(input)
    input = lowercase(input)
    d = Dict{SubString{typeof(input)}, Int}()
    iswordchar(c) = !ispunct(c) && !isspace(c)

    isempty(input) && return d

    in_word = false
    start_of_word = 0
    prev = 1
    idx = 1
    while true
        chr = input[idx]
        if iswordchar(chr)
            if !in_word
                in_word = true
                start_of_word = idx
            end
        elseif in_word
            # We've reached the end of a word
            in_word = false
            word = SubString(input, start_of_word:prev)
            d[word] = get(d, word, 0) + 1
        end
        if idx == lastindex(input)
            break
        end
        prev = idx
        idx = nextind(input, idx)
    end
    # Did input end on a word?
    if in_word
        word = SubString(input, start_of_word:idx)
        d[word] = get(d, word, 0) + 1
    end
    return d
end
```

Dunno if you prefer this. I don't mind.

Reply:
I don't know if LICM will hoist `lastindex` out, btw. If it doesn't then this does a load of unnecessary work.

Reply:
Wow, I just assumed the lastindex check would be hoisted and didn't even bother to check! Precomputing it is a full 10% faster. 
The above version though didn't work for me on contractions. I thought only the lookahead was missing, but it turned out more involved since adding it naively introduced a boundserror. The current working version, uncommented looks like this (it's a shame syntax highlighting is better in comments than in the markdown preview...)

```julia
function wordcount(input)
    input = lowercase(input)
    d = Dict{SubString{typeof(input)}, Int}()
    iswordchar(c) = !ispunct(c) && !isspace(c)

    in_word = false
    prev = 0
    lastind = lastindex(input)
    for i in eachindex(input)
        c = input[i]
        if i == lastind
            prev = in_word ? prev : i
            i = iswordchar(c) ? i : i-1
            if prev <= i
                k = SubString(input, prev:i)
                d[k] = get(d, k, 0) + 1
            end
        elseif iswordchar(c)
            if !in_word
                in_word = true
                prev = i
            end
        elseif in_word
            c == ''' && iswordchar(input[nextind(input, i)]) && continue

            in_word = false
            k = SubString(input, prev:i-1)
            d[k] = get(d, k, 0) + 1
        end
    end
    d
end
```

Reply:
Lol, I should stop posting up quick, untested code.

Here's a version that works, but I think your one is better.

```jl
function wordcount(input)
    input = lowercase(input)
    d = Dict{SubString{typeof(input)}, Int}()
    iswordchar(c) = !ispunct(c) && !isspace(c)

    isempty(input) && return d

    in_word = false
    start_of_word = 0
    prev = 1
    idx = 1
    while true
        chr = input[idx]
        if iswordchar(chr)
            if !in_word
                in_word = true
                start_of_word = idx
            end
        elseif in_word
            # We've reached the end of a word
            in_word = false
            word = SubString(input, start_of_word:prev)
            d[word] = get(d, word, 0) + 1
        end
        if idx == lastindex(input)
            break
        end
        prev = idx
        idx = nextind(input, idx)
    end
    # Did input end on a word?
    if in_word
        word = SubString(input, start_of_word:idx)
        d[word] = get(d, word, 0) + 1
    end
    return d
end
```

Reply:
And yeah, I think Julia is currently not very compatible with LLVM's LICM pass, or possibly with some other LLVM pass, and lots of stuff that one would like hoisted isn't. I think the Julia compiler people are aware and there's some work that might fix it in the future.

e.g. anything that can error usually can't be hoisted, and while lastindex won't error on valid Strings, it might error on strings that contain bytes that are illegal in UTF-8, for example.